Tests/add product

GuilhermeMota93 opened the pull request:
## RESOURCES
[The Figma file](https://www.figma.com/file/QAkdm5HZdLlG0b8vUkv1pX/Products---Add-%26-Edit-Products?node-id=460%3A1760&viewport=-1704%2C1577%2C0.5334523320198059) and [the initial issue](https://github.com/woocommerce/woocommerce-android/issues/2707)

## CHANGES / DESCRIPTION

This PR will contain the following:
- added tests for the **add product** view model flow
- added tests for the product list 
- refactored `AppPrefs` (a little bit!) to allow being injected in different places - **helps going forward with tests** 👍 

Starting from the last bit, refactored the way `AppPrefs` is used a bit by having it injected in different view models and also in the main `WooCommerce` application class. Basically, to achieve this, I added the following in the `AppConfigModule`:
```
@Provides
fun providesAppPrefs(appContext: Context): AppPrefs {
    AppPrefs.init(appContext)
    return AppPrefs
}
```
This means that it can now be injected 🎉 

I then updated the `WooCommerce` to put things in the correct order and use the injected preferences class. As far as I am aware, this should be fine, even calling `AppPrefs` from different places and directly (aka, not injected) as the `init` already happened and the context is the same from the start. 

I double checked the app running and checked the logs for analytics and such, and it all seems to be working fine...but **it needs proper testing to make sure nothing broke** 🙏 

Now for the new tests!

Added a new test scenario named `ProductDetailViewModel_AddFlowTest` (not a great naming, but it's descriptive) that will handle the **add product** test flow. It's injected with the correct `navArgs` and uses a mock `AppPrefs` in order to get the correct selected default `type` of product 👍 

I also added a test on the `ProductListViewModelTest` for the check of the **add button** visibility given the different loading scenarios.

As always, let me know your thoughts 🙏 


**EDIT**
Also:
- added the `PUBLISH` to `UPDATE` logic in order to display the top nav bar navigation and correct loading/success dialogs, depending on the flow

## SCREENSHOTS
All the tests
-
<img width="500" alt="Screenshot 2020-08-28 at 00 15 28" src="https://user-images.githubusercontent.com/7058393/91503905-98ea9a80-e8c3-11ea-9afb-9274e1107fc0.png">

Specific ones modified
-

<img width="300" alt="Screenshot 2020-08-28 at 12 44 13" src="https://user-images.githubusercontent.com/7058393/91557393-4f826580-e92c-11ea-90d5-a24b4d6c1248.png">
<img width="300" alt="Screenshot 2020-08-28 at 00 11 57" src="https://user-images.githubusercontent.com/7058393/91503724-27aae780-e8c3-11ea-879b-6014ef010699.png">
<img width="300" alt="Screenshot 2020-08-28 at 00 12 09" src="https://user-images.githubusercontent.com/7058393/91503725-28437e00-e8c3-11ea-9005-d760f769fb47.png">

Add Product flow to Update flow
-
<image src="https://user-images.githubusercontent.com/7058393/91557005-96bc2680-e92b-11ea-8c66-37e0137ad014.gif" width="350"/>

Commit:
Added product list view model test to assert visibility of the add new product button

Commit:
Refactored AppPrefs to be injected via Dagger - useful for testing - into the main app class and the view models where it's needed. Added tests for the product detail view model on the add flow and the product list view model.

GuilhermeMota93 commented on WooCommerce/src/main/kotlin/com/woocommerce/android/WooCommerce.kt:
```diff
@@ -119,17 +121,19 @@ open class WooCommerce : MultiDexApplication(), HasAndroidInjector, ApplicationL
+        val wellSqlConfig = WooWellSqlConfig(applicationContext)
+        WellSql.init(wellSqlConfig)
+
+        component.inject(this)
```
Moved these around to allow proper initialisation of components, in the needed order 👍

Commit:
Merge branch 'issue/2707-add-simple-products' into tests/add-product

* issue/2707-add-simple-products:
  Added the default values for stock and catalog visibility in order to fix remote call error. Added the correct logic to either trigger an add product or update existing product in the add flow.

0nko commented on WooCommerce/src/test/kotlin/com/woocommerce/android/ui/products/ProductDetailViewModel_AddFlowTest.kt:
```diff
@@ -0,0 +1,279 @@
+        assertThat(cards).isEqualTo(expectedCards)
+    }
+
+    private fun stripCallbacks(card: ProductPropertyCard): ProductPropertyCard {
```
Since the same method is used in different places now, we could move it to a common util class so that it can be reused.

0nko reviewed (changes requested):
This looks good! 🎉 

One thing that must be fixed is detecting if we're in **add** or **edit** flow. You're already doing that in a couple of places, but the progress dialog message or Publish button still show the **add** state even after a product has been published. You could add a property with a value `navArgs.isAddProduct && viewState.productDraft?.remoteId == 0L` and just use that.
 
You could also add a test for this that makes sure that the proper flow is used after a product is published.

Also, `ktlint` is failing.

Commit:
Updated dialog display depending on flow edit/add. Updated the top nav bar menu done title depending on edit/add flow.

Commit:
Added test to validate correct update flow to be triggered when updating a freshly added product

Commit:
Tidy up for lint

0nko commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/ProductDetailFragment.kt:
```diff
@@ -238,11 +245,20 @@ class ProductDetailFragment : BaseProductFragment(), OnGalleryImageClickListener
 
     private fun setupProductAddOptionsMenu(menu: Menu) {
         menu.findItem(R.id.menu_done).apply {
```
You can use the `doneOrUpdateMenuItem` here.

0nko reviewed (changes requested):
When you publish a product and rotate the screen with **Don't keep activites** enabled, the done button's title is set to "PUBLISH" again [here](https://github.com/woocommerce/woocommerce-android/pull/2798/files#diff-2c2eab32dd2dee7f590516c59e7b6f56L232).

![publish](https://user-images.githubusercontent.com/1522856/91562090-e9520e80-e93c-11ea-995a-98461e7bccb9.gif)

It might be easier to just create a property and use that so it's not duplicated everywhere and you won't have to worry about forgetting about the 2nd condition:
```
val isAddFlow: Boolean 
    get() = navArgs.isAddProduct && remoteId == ProductDetailViewModel.DEFAULT_ADD_NEW_PRODUCT_ID
```

What do you think?

Commit:
Tidied up code by introducing isAddFlow and isAddFlowEntryPoint helpers in the product detail view model

GuilhermeMota93 commented:
@0nko updated as per your comments, very good shout 👍 it should be fixed now.

There was a slight split in the checks though, as one needs to check at the `start` of the view model if the entry point is for `navArgs.isAddProduct` without considering the product id - as at this point is **null** from the product draft

I introduced two helpers in the detail view model to help with it:
- `isAddFlowEntryPoint`
- `isAddFlow`

These should be quite descriptive as I also added some explanatory business logic in each helper 👌 let me know what you think!

Add new Product to rotate and Update
-
<image src="https://user-images.githubusercontent.com/7058393/91579384-6ed8bd00-e943-11ea-8e9e-ed6e0e0e0aca.png" width="450"/>

0nko reviewed (approved):
Thanks for the fixes @GuilhermeMota93! 👍 

:shipit:
